File: src/com/gdb/domain/AbstractAccount.java
package com.gdb.domain;
import com.gdb.exceptions.*;

public abstract class AbstractAccount {
    protected String accountNumber;
    protected String name;
    protected int age;
    protected double balance;
    protected String accountType;
    protected String status;
    protected String pin;

    public AbstractAccount(String accountNumber, String name, int age, double balance, String accountType, String status, String pin) {
        if (age < 18) throw new IllegalArgumentException("Customer age must be at least 18.");
        if (balance < 0) throw new IllegalArgumentException("Initial balance cannot be negative.");
        if (pin == null || pin.length() != 4 || !pin.matches("\\d{4}")) throw new IllegalArgumentException("PIN must be exactly 4 digits.");
        
        this.accountNumber = accountNumber;
        this.name = name;
        this.age = age;
        this.balance = balance;
        this.accountType = accountType;
        this.status = status;
        this.pin = pin;
    }

    public boolean deposit(double amount) {
        if (amount > 0) {
            this.balance += amount;
            return true;
        }
        return false;
    }

    public boolean validatePin(String inputPin) {
        return this.pin != null && this.pin.equals(inputPin);
    }

    public boolean changePin(String oldPin, String newPin) {
        if (validatePin(oldPin)) {
            if (newPin != null && newPin.length() == 4 && newPin.matches("\\d{4}")) {
                this.pin = newPin;
                return true;
            }
        }
        return false;
    }

    public void displayAccountInfo() {
        System.out.println("Account Number: " + accountNumber);
        System.out.println("Customer Name: " + name);
        System.out.println("Customer Age: " + age);
        System.out.println("Account Balance: " + balance);
        System.out.println("Account Type: " + accountType);
        System.out.println("Account Status: " + status);
    }

    public final void withdraw(double amount, String enteredPin) throws AccountException {
        if (!validatePin(enteredPin)) {
            throw new InvalidPinException("Invalid PIN entered");
        }
        if (!"ACTIVE".equals(status)) {
            throw new InactiveAccountException("Account is not active");
        }
        if (amount <= 0) {
            throw new InvalidAmountException("Withdrawal amount must be positive");
        }
        processDebit(amount);
    }

    protected abstract void processDebit(double amount) throws AccountException;

    public void suspend() { this.status = "SUSPENDED"; }
    public void activate() { this.status = "ACTIVE"; }
    public void close() { this.status = "CLOSED"; }
    public double getBalance() { return balance; }
}

File: gdb/domain/abstract_account.py

from abc import abstractmethod
from gdb.domain.iaccount import IAccount
from gdb.exceptions import *
from gdb.domain.account_rules_engine import AccountRulesEngine

class AbstractAccount(IAccount):
    def __init__(self, account_number, name, age, balance, account_type, status, pin):
        if age < 18: raise ValueError("Customer age must be at least 18.")
        if balance < 0: raise ValueError("Initial balance cannot be negative.")
        if not pin or len(pin) != 4 or not pin.isdigit(): raise ValueError("PIN must be exactly 4 digits.")
        
        self.account_number = account_number
        self.name = name
        self.age = age
        self.balance = balance
        self.account_type = account_type
        self.status = status
        self.pin = pin
        self.rules_engine = AccountRulesEngine()

    def deposit(self, amount: float) -> bool:
        if amount > 0:
            self.balance += amount
            return True
        return False

    def validate_pin(self, input_pin: str) -> bool:
        return self.pin == input_pin

    def withdraw(self, amount: float, entered_pin: str) -> None:
        if not self.validate_pin(entered_pin):
            raise InvalidPinException("Invalid PIN entered")
        if self.status != "ACTIVE":
            raise InactiveAccountException("Account is not active")
        if amount <= 0:
            raise InvalidAmountException("Withdrawal amount must be positive")
        
        self.rules_engine.validate_withdrawal(self, amount)
        self.process_debit(amount)

    @abstractmethod
    def process_debit(self, amount: float) -> None:
        pass

    def get_balance(self) -> float:
        return self.balance
        
    def display_account_info(self) -> None:
        pass
        
    def suspend(self) -> None: self.status = "SUSPENDED"
    def activate(self) -> None: self.status = "ACTIVE"
    def close(self) -> None: self.status = "CLOSED"
